### ISSUE TITLE
Vernac controls and focusing bracket in CoqIDE

### ISSUE THREAD

**Issue by Zimmi48** [coq/coq#6587]
#### Version

8.7.1 and 8.8+alpha (3815ec540)

#### Operating system

Linux / NixOS

#### Description of the problem

The following script:

```
Goal True.
Time { split. }
Qed.
```

is correctly processed by coqc. It also passes in Proof-General with a printing bug (the goal is still printed after it has been solved by `split`) but it doesn't pass in CoqIDE.

When evaluating step by step in CoqIDE we get:

```
This proof is focused, but cannot be unfocused this way
```

Basically the `split` was ignored. (BTW the error message is missing a dot.)

When evaluating the whole buffer, it fails at the same place but with an anomaly `Uncaught exception Stm.Vcs_aux.Expired`.

Tagging @maximedenes for the vernac control part, and @ejgallego and @gares for the STM anomaly.

**Comment by psteckler** [comment]
The error happens when using the XML protocol, so it fails in the same way with PG/xml.

Here's the relevant XML traffic, as recorded in PG/xml:
```
*emacs:
<call val="Add"><pair><pair><string>Time&nbsp;{&nbsp;split.</string>
<int>-3</int>
</pair>
<pair><state_id val="3"/><bool val="true"/></pair>
</pair>
</call>

*coqtop:
<value val="good"><pair><state_id val="4"/><pair><union val="in_l"><unit/></union><string></string></pair></pair></value>

*emacs:
<call val="Goal"><unit/></call>
...
*coqtop:
<value val="good"><option val="some"><goals><list><goal><string>1</string><list/><richpp><_><pp><constr.reference>True</constr.reference></pp></_></richpp></goal></list><list><pair><list/><list/></pair></list><list/><list/></goals></option></value>

...
*emacs:
<call val="Add"><pair><pair><string>}</string>
<int>-4</int>
</pair>
<pair><state_id val="4"/><bool val="true"/></pair>
</pair>
</call>

*coqtop:
<value val="good"><pair><state_id val="5"/><pair><union val="in_l"><unit/></union><string></string></pair></pair></value>

*emacs:
<call val="Goal"><unit/></call>
...
*coqtop:
<value val="fail" loc_s="0" loc_e="1"><state_id val="4"/><richpp><_><pp>This&nbsp;proof&nbsp;is&nbsp;focused,&nbsp;but&nbsp;cannot&nbsp;be&nbsp;unfocused&nbsp;this&nbsp;way</pp></_></richpp></value>
```

In vanilla PG, we see:
```
<prompt>Unnamed_thm < 7 |branch| 0 < </prompt>Time { split.
<infomsg>Finished transaction in 0. secs (0.u,0.s) (successful)</infomsg>
1 subgoal (ID 1)
  
  ============================
  True

<prompt>Unnamed_thm < 8 |branch| 0 < </prompt>No more subgoals.

<prompt>Unnamed_thm < 9 |branch| 0 < </prompt>}
No more subgoals.
```
Note that the goal has been discharged before the terminating right-bracket has been sent, while in the XML above, the `True` goal remains at that point.

**Comment by ejgallego** [comment]
I'm much afraid that in @psteckler 's trace the bug is in emacs not in the protocol. The protocol is parsing the string correctly, but Emacs is not sending the right tokenization thus it obviously fails. You can see it better with more recent protocols that admit batch addition:
```
(Add () "Goal True. Time { split. } Qed.")
(Answer 0 Ack)
(Answer 0(Added 2((fname"")(line_nb 1)(bol_pos 0)(line_nb_last 1)(bol_pos_last 0)(bp 0)(ep 10))NewTip))
(Answer 0(Added 3((fname"")(line_nb 1)(bol_pos 0)(line_nb_last 1)(bol_pos_last 0)(bp 11)(ep 17))NewTip))
(Answer 0(Added 4((fname"")(line_nb 1)(bol_pos 0)(line_nb_last 1)(bol_pos_last 0)(bp 18)(ep 24))NewTip))
(Answer 0(Added 5((fname"")(line_nb 1)(bol_pos 0)(line_nb_last 1)(bol_pos_last 0)(bp 25)(ep 26))NewTip))
(Answer 0(Added 6((fname"")(line_nb 1)(bol_pos 0)(line_nb_last 1)(bol_pos_last 0)(bp 27)(ep 31))NewTip))
(Answer 0 Completed)
```

The input is parsed and split correctly by Coq.

**Comment by psteckler** [comment]
The tokenization is the same, of course, in PG and PG/xml, which share that code.

**Comment by psteckler** [comment]
@ejgallego So the response indicates 5 tokens, which should be:
 - the goal statement
 - Time with left bracket
 - split statement
 - right bracket
 - Qed

?

**Comment by psteckler** [comment]
If I leave out the `Time`, PG/xml and CoqIDE both do the right thing.

**Comment by psteckler** [comment]
I've created the issue ProofGeneral/PG#223.

**Comment by ejgallego** [comment]
Yup that's the tokenization the Coq parser expects.

**Comment by psteckler** [comment]
A grouch's remark: it's kind-of-awful that the grammar of Coq is specified separately for each of Coq, CoqIDE, and PG.

**Comment by ejgallego** [comment]
Use a protocol that doesn't require the IDE to know the grammar. [Hint: the grammar is user-modifiable]

**Comment by ejgallego** [comment]
Or dump the grammar and have your IDE understand it XD

**Comment by psteckler** [comment]
@ejgallego Yes, those would be excellent things to do. :-)

**Comment by psteckler** [comment]
I created a PR that seems to fix the issue in PG: ProofGeneral/PG#224.

**Comment by psteckler** [comment]
For CoqIDE, I created: https://github.com/coq/coq/pull/6652

**Comment by psteckler** [comment]
Do we still want to fix the anomaly here? Should bad input be allowed to trigger anomalies?

**Comment by Zimmi48** [comment]
No anomalies should always be considered bugs. Bad input should be caught and emit an error.

**Comment by Zimmi48** [comment]
As noted in the PR thread, `Time` is not the only vernacular control and thus not the only one to require fixing on this example (there are also `Timeout`, `Redirect` and `Fail`). Moreover, after #6551, the wrong behavior can also be observed with:

```
Goal True.
Time 1: { fail. }
Qed.
```

It would be good if we could have a generic solution here.

**Comment by psteckler** [comment]
I added support for the other known controls in the PR for CoqIDE.

I will leave the anomaly fix to STM experts.

**Comment by ejgallego** [comment]
I didn't check it but it looks like the anomaly is just coming from a call of the client referring to a state that is not valid.

**Comment by psteckler** [comment]
Fixed in PG and PG/xml, by the way.

**Comment by Zimmi48** [comment]
I think we can close this now. The anomaly cannot be reproduced in master after #6652.

**Comment by psteckler** [comment]
@Zimmi48 Note that while the anomaly no longer occurs if an IDE submits valid statements, which CoqIDE and PG do now, an incorrectly-tokenized script will still cause an anomaly.

**Comment by Zimmi48** [comment]
Ah OK. Maybe a new issue should be opened with a reproducible test case using `cat` to pipe into `coqtop -ideslave` or something like this. While still a bug, anomalies triggered when incorrectly using the IDE protocol should be considered lower priority than anomalies that appear directly to the user.

**Comment by ejgallego** [comment]
@Zimmi48 I recommend not caring about the anomalies, they should be eliminated as the STM/protocol roadmap gets gradually implemented.
